In [47]:
# import needed libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re

import warnings
warnings.filterwarnings('ignore')

In [48]:
#Read the data file
telecom = pd.read_csv("train.csv")
telecom.head()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0   0        109             0.0             0.0             0.0   
1   1        109             0.0             0.0             0.0   
2   2        109             0.0             0.0             0.0   
3   3        109             0.0             0.0             0.0   
4   4        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
0            6/30/2014            7/31/2014            8/31/2014   31.277   
1            6/30/2014            7/31/2014            8/31/2014    0.000   
2            6/30/2014            7/31/2014            8/31/2014   60.806   
3            6/30/2014            7/31/2014            8/31/2014  156.362   
4            6/30/2014            7/31/2014            8/31/2014  240.708   

    arpu_7  ...  sachet_3g_7  sachet_3g_8  fb_user_6  fb_user_7  fb_user_8  \
0   87.009  ...            0            0        NaN        NaN        NaN   
1  122.787  ...            0            0        NaN        1.0        NaN   
2  103.176  ...            0            0        NaN        NaN        NaN   
3  205.260  ...            0            0        NaN        NaN        NaN   
4  128.191  ...            1            0        1.0        1.0        1.0   

    aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  churn_probability  
0  1958         0.0         0.0         0.0                  0  
1   710         0.0         0.0         0.0                  0  
2   882         0.0         0.0         0.0                  0  
3   982         0.0         0.0         0.0                  0  
4   647         0.0         0.0         0.0                  0  

[5 rows x 172 columns]

In [49]:
#Check for data dimension
telecom.shape

(69999, 172)

In [50]:
telecom.info(verbose=1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 172 columns):
 #    Column                    Dtype  
---   ------                    -----  
 0    id                        int64  
 1    circle_id                 int64  
 2    loc_og_t2o_mou            float64
 3    std_og_t2o_mou            float64
 4    loc_ic_t2o_mou            float64
 5    last_date_of_month_6      object 
 6    last_date_of_month_7      object 
 7    last_date_of_month_8      object 
 8    arpu_6                    float64
 9    arpu_7                    float64
 10   arpu_8                    float64
 11   onnet_mou_6               float64
 12   onnet_mou_7               float64
 13   onnet_mou_8               float64
 14   offnet_mou_6              float64
 15   offnet_mou_7              float64
 16   offnet_mou_8              float64
 17   roam_ic_mou_6             float64
 18   roam_ic_mou_7             float64
 19   roam_ic_mou_8             float64
 20   roam

Original data file consists of 69999 rows and 172 columns. As per the results there are both numerical and categorical columns

## Handling missing values

In [51]:
#Checking for further missing values per column
null_value_series = telecom.isnull().sum()/len(telecom.index)*100

#Listing columns having more than 5% missing values
null_value_series = null_value_series[null_value_series > 5]
null_value_series

onnet_mou_8                  5.290076
offnet_mou_8                 5.290076
roam_ic_mou_8                5.290076
roam_og_mou_8                5.290076
loc_og_t2t_mou_8             5.290076
loc_og_t2m_mou_8             5.290076
loc_og_t2f_mou_8             5.290076
loc_og_t2c_mou_8             5.290076
loc_og_mou_8                 5.290076
std_og_t2t_mou_8             5.290076
std_og_t2m_mou_8             5.290076
std_og_t2f_mou_8             5.290076
std_og_t2c_mou_8             5.290076
std_og_mou_8                 5.290076
isd_og_mou_8                 5.290076
spl_og_mou_8                 5.290076
og_others_8                  5.290076
loc_ic_t2t_mou_8             5.290076
loc_ic_t2m_mou_8             5.290076
loc_ic_t2f_mou_8             5.290076
loc_ic_mou_8                 5.290076
std_ic_t2t_mou_8             5.290076
std_ic_t2m_mou_8             5.290076
std_ic_t2f_mou_8             5.290076
std_ic_t2o_mou_8             5.290076
std_ic_mou_8                 5.290076
spl_ic_mou_8

#### There are many columns having > 70 % missing values but zero columns are having missing values

### In context of business its important to keep values for rechange columns and hence imputing zero for all rechange columns

In [52]:
recharge_columns = ['total_rech_data_6', 'total_rech_data_7', 'total_rech_data_8',
                 'max_rech_data_6', 'max_rech_data_7', 'max_rech_data_8',
                 'av_rech_amt_data_6', 'av_rech_amt_data_7', 'av_rech_amt_data_8'
                 ]

In [53]:
# impute missing values with 0
telecom[recharge_columns] = telecom[recharge_columns].apply(lambda x: x.fillna(0))

In [54]:
# now, let's make sure values are imputed correctly
print("Missing value ratio:\n")
print(telecom[recharge_columns].isnull().sum()*100/telecom.shape[1])

# summary
print("\n\nSummary statistics\n")
print(telecom[recharge_columns].describe(include='all'))

Missing value ratio:

total_rech_data_6     0.0
total_rech_data_7     0.0
total_rech_data_8     0.0
max_rech_data_6       0.0
max_rech_data_7       0.0
max_rech_data_8       0.0
av_rech_amt_data_6    0.0
av_rech_amt_data_7    0.0
av_rech_amt_data_8    0.0
dtype: float64


Summary statistics

       total_rech_data_6  total_rech_data_7  total_rech_data_8  \
count       69999.000000       69999.000000       69999.000000   
mean            0.619309           0.683981           0.697867   
std             1.762010           1.943193           1.973434   
min             0.000000           0.000000           0.000000   
25%             0.000000           0.000000           0.000000   
50%             0.000000           0.000000           0.000000   
75%             1.000000           1.000000           1.000000   
max            61.000000          54.000000          60.000000   

       max_rech_data_6  max_rech_data_7  max_rech_data_8  av_rech_amt_data_6  \
count     69999.000000     69999

### Checking for columns having single value across all rows

In [55]:
#Checking for columns having majority of single values for the dataset and percentage of single value in each column
telecom.nunique().sort_values().head(25)

std_ic_t2o_mou_6         1
circle_id                1
loc_og_t2o_mou           1
std_og_t2o_mou           1
loc_ic_t2o_mou           1
last_date_of_month_6     1
last_date_of_month_7     1
last_date_of_month_8     1
std_ic_t2o_mou_8         1
std_og_t2c_mou_8         1
std_og_t2c_mou_7         1
std_og_t2c_mou_6         1
std_ic_t2o_mou_7         1
night_pck_user_6         2
night_pck_user_7         2
night_pck_user_8         2
churn_probability        2
fb_user_7                2
fb_user_8                2
fb_user_6                2
monthly_2g_6             5
monthly_2g_8             6
monthly_2g_7             6
monthly_3g_6            10
monthly_3g_8            12
dtype: int64

#### Many columns are having single values present for all the rows and hence dropping these columns as these columns won't be helpful in determining Chrun probability

In [56]:
#Create Data frame for Features and value counts
telecom_count_df = pd.DataFrame(telecom.nunique().sort_values()).reset_index()
telecom_count_df.columns = ['Features','Value_count']

#Get name of columns having Value_count = 1
single_value_col = telecom_count_df[telecom_count_df['Value_count'] == 1]['Features'].tolist()
single_value_col

['std_ic_t2o_mou_6',
 'circle_id',
 'loc_og_t2o_mou',
 'std_og_t2o_mou',
 'loc_ic_t2o_mou',
 'last_date_of_month_6',
 'last_date_of_month_7',
 'last_date_of_month_8',
 'std_ic_t2o_mou_8',
 'std_og_t2c_mou_8',
 'std_og_t2c_mou_7',
 'std_og_t2c_mou_6',
 'std_ic_t2o_mou_7']

In [57]:
print("Shape before dropping single value columns : ", telecom.shape)
#Delete single value columns
telecom = telecom.drop(single_value_col,axis=1)
print("Shape after dropping single value columns : ", telecom.shape)

Shape before dropping single value columns :  (69999, 172)
Shape after dropping single value columns :  (69999, 159)


### Check the distribution of values for columns having 2 unique values

In [58]:
telecom.night_pck_user_6.value_counts(normalize=True)*100


0.0    97.472678
1.0     2.527322
Name: night_pck_user_6, dtype: float64

In [59]:
telecom.night_pck_user_7.value_counts(normalize=True)*100

0.0    97.593059
1.0     2.406941
Name: night_pck_user_7, dtype: float64

In [60]:
telecom.night_pck_user_8.value_counts(normalize=True)*100

0.0    97.898681
1.0     2.101319
Name: night_pck_user_8, dtype: float64

In [61]:
telecom.fb_user_6.value_counts(normalize=True)*100

1.0    91.632514
0.0     8.367486
Name: fb_user_6, dtype: float64

In [62]:
telecom.fb_user_7.value_counts(normalize=True)*100

1.0    90.95438
0.0     9.04562
Name: fb_user_7, dtype: float64

In [63]:
telecom.fb_user_8.value_counts(normalize=True)*100

1.0    89.031873
0.0    10.968127
Name: fb_user_8, dtype: float64

### Dropping columns having more than 93% of single values


In [64]:
print("Shape before dropping 93% single value columns : ", telecom.shape)
telecom = telecom.drop(['night_pck_user_6','night_pck_user_7','night_pck_user_8'],axis=1)
print("Shape after dropping 93% single value columns : ", telecom.shape)

Shape before dropping 93% single value columns :  (69999, 159)
Shape after dropping 93% single value columns :  (69999, 156)


### Computing categorical and numeric columns

In [65]:
#Numeric
numeric_col = list(telecom.dtypes[telecom.dtypes != 'object'].index)
print((numeric_col))

['id', 'arpu_6', 'arpu_7', 'arpu_8', 'onnet_mou_6', 'onnet_mou_7', 'onnet_mou_8', 'offnet_mou_6', 'offnet_mou_7', 'offnet_mou_8', 'roam_ic_mou_6', 'roam_ic_mou_7', 'roam_ic_mou_8', 'roam_og_mou_6', 'roam_og_mou_7', 'roam_og_mou_8', 'loc_og_t2t_mou_6', 'loc_og_t2t_mou_7', 'loc_og_t2t_mou_8', 'loc_og_t2m_mou_6', 'loc_og_t2m_mou_7', 'loc_og_t2m_mou_8', 'loc_og_t2f_mou_6', 'loc_og_t2f_mou_7', 'loc_og_t2f_mou_8', 'loc_og_t2c_mou_6', 'loc_og_t2c_mou_7', 'loc_og_t2c_mou_8', 'loc_og_mou_6', 'loc_og_mou_7', 'loc_og_mou_8', 'std_og_t2t_mou_6', 'std_og_t2t_mou_7', 'std_og_t2t_mou_8', 'std_og_t2m_mou_6', 'std_og_t2m_mou_7', 'std_og_t2m_mou_8', 'std_og_t2f_mou_6', 'std_og_t2f_mou_7', 'std_og_t2f_mou_8', 'std_og_mou_6', 'std_og_mou_7', 'std_og_mou_8', 'isd_og_mou_6', 'isd_og_mou_7', 'isd_og_mou_8', 'spl_og_mou_6', 'spl_og_mou_7', 'spl_og_mou_8', 'og_others_6', 'og_others_7', 'og_others_8', 'total_og_mou_6', 'total_og_mou_7', 'total_og_mou_8', 'loc_ic_t2t_mou_6', 'loc_ic_t2t_mou_7', 'loc_ic_t2t_mou_8

In [66]:
cat_col = list(telecom.dtypes[telecom.dtypes == 'object'].index)
print((cat_col))

['date_of_last_rech_6', 'date_of_last_rech_7', 'date_of_last_rech_8', 'date_of_last_rech_data_6', 'date_of_last_rech_data_7', 'date_of_last_rech_data_8']


In [67]:
#From above data ID column from numeric type and Date columns from categorical types can be dropped
telecom = telecom.drop(cat_col,axis=1)
telecom = telecom.drop(['id'],axis=1)
print("Shape after dropping ID and date columns ", telecom.shape)


Shape after dropping ID and date columns  (69999, 149)


### Drop variables with more than a given threshold of missing values

In [68]:
# Below code gives percentage of null in every column
null_percentage = telecom.isnull().sum()/telecom.shape[0]*100

# Below code gives list of columns having more than 60% null
col_to_drop = null_percentage[null_percentage>70].keys()

intial_cols = telecom.shape[1]
telecom = telecom.drop(col_to_drop, axis=1)

In [69]:
print("Columns dropped due to 70% missing values \n")
print("\n".join(col_to_drop.to_list()))
print("\n{0} columns dropped.".format(intial_cols-telecom.shape[1]))

Columns dropped due to 70% missing values 

count_rech_2g_6
count_rech_2g_7
count_rech_2g_8
count_rech_3g_6
count_rech_3g_7
count_rech_3g_8
arpu_3g_6
arpu_3g_7
arpu_3g_8
arpu_2g_6
arpu_2g_7
arpu_2g_8
fb_user_6
fb_user_7
fb_user_8

15 columns dropped.


### Impute with mice for other missing values

In [70]:
# using MICE technique to impute missing values in the rest of the columns
from fancyimpute import IterativeImputer as MICE
telecom_imputed = MICE().fit_transform(telecom)

In [71]:
# convert imputed numpy array to pandas dataframe
telecom = pd.DataFrame(telecom_imputed, columns=telecom.columns)
print(telecom.isnull().sum()*100/telecom.shape[0])

arpu_6               0.0
arpu_7               0.0
arpu_8               0.0
onnet_mou_6          0.0
onnet_mou_7          0.0
                    ... 
aon                  0.0
aug_vbc_3g           0.0
jul_vbc_3g           0.0
jun_vbc_3g           0.0
churn_probability    0.0
Length: 134, dtype: float64


Data is cleaned with zero missing values

### Derive high value customers
1. From the data set we noticed that if data and recharge value is decreasing in Aug month compared to previous months, such customer are more likely to churn
2. It is important to retain high value customers and determine Churn rate for such high value customer
3. High value customers are calculated based on June and July month recharges

In [72]:
# calculate total recharge -->> call recharge amount + data recharge amount
telecom['total_rech_6'] = telecom.total_rech_amt_6 + telecom.av_rech_amt_data_6
telecom['total_rech_7'] = telecom.total_rech_amt_7 + telecom.av_rech_amt_data_7

In [73]:
# calculate average recharge(data + recharge) done by customer in June and July
telecom['av_rech_amt_data_6_7'] = (telecom.total_rech_6 + telecom.total_rech_7)/2

In [74]:
# look at the 70th percentile recharge amount
print("Recharge amount at 70th percentile: {0}".format(telecom.av_rech_amt_data_6_7.quantile(0.7)))

Recharge amount at 70th percentile: 431.0


In [75]:
# retain only those customers who have recharged their mobiles with more than or equal to 70th percentile amount
telecom_filtered = telecom.loc[telecom.av_rech_amt_data_6_7 >= telecom.av_rech_amt_data_6_7.quantile(0.7), :]
telecom_filtered = telecom_filtered.reset_index(drop=True)
telecom_filtered.shape

(21041, 137)

In [76]:
# delete variables created to filter high-value customers
#telecom_filtered = telecom_filtered.drop(['av_rech_amt_data_6', 'av_rech_amt_data_7',
 #                                     'total_rech_6', 'total_rech_7', 'av_rech_amt_data_6_7'], axis=1)
#telecom_filtered.shape

We are having 21041 high value customers after filtering based on 70th percentile recharge amount

## Calculate difference between 8th and previous months
The most important feature, in this situation, can be the difference between the 8th month and the previous months. The difference can be in patterns such as usage difference or recharge value difference. Let's calculate difference variable as the difference between 8th month and the average of 6th and 7th

In [77]:
telecom_filtered['arpu_diff'] = telecom_filtered.arpu_8 - ((telecom_filtered.arpu_6 + telecom_filtered.arpu_7)/2)

telecom_filtered['onnet_mou_diff'] = telecom_filtered.onnet_mou_8 - ((telecom_filtered.onnet_mou_6 + telecom_filtered.onnet_mou_7)/2)

telecom_filtered['offnet_mou_diff'] = telecom_filtered.offnet_mou_8 - ((telecom_filtered.offnet_mou_6 + telecom_filtered.offnet_mou_7)/2)

telecom_filtered['roam_ic_mou_diff'] = telecom_filtered.roam_ic_mou_8 - ((telecom_filtered.roam_ic_mou_6 + telecom_filtered.roam_ic_mou_7)/2)

telecom_filtered['roam_og_mou_diff'] = telecom_filtered.roam_og_mou_8 - ((telecom_filtered.roam_og_mou_6 + telecom_filtered.roam_og_mou_7)/2)

telecom_filtered['loc_og_mou_diff'] = telecom_filtered.loc_og_mou_8 - ((telecom_filtered.loc_og_mou_6 + telecom_filtered.loc_og_mou_7)/2)

telecom_filtered['std_og_mou_diff'] = telecom_filtered.std_og_mou_8 - ((telecom_filtered.std_og_mou_6 + telecom_filtered.std_og_mou_7)/2)

telecom_filtered['isd_og_mou_diff'] = telecom_filtered.isd_og_mou_8 - ((telecom_filtered.isd_og_mou_6 + telecom_filtered.isd_og_mou_7)/2)

telecom_filtered['spl_og_mou_diff'] = telecom_filtered.spl_og_mou_8 - ((telecom_filtered.spl_og_mou_6 + telecom_filtered.spl_og_mou_7)/2)

telecom_filtered['total_og_mou_diff'] = telecom_filtered.total_og_mou_8 - ((telecom_filtered.total_og_mou_6 + telecom_filtered.total_og_mou_7)/2)

telecom_filtered['loc_ic_mou_diff'] = telecom_filtered.loc_ic_mou_8 - ((telecom_filtered.loc_ic_mou_6 + telecom_filtered.loc_ic_mou_7)/2)

telecom_filtered['std_ic_mou_diff'] = telecom_filtered.std_ic_mou_8 - ((telecom_filtered.std_ic_mou_6 + telecom_filtered.std_ic_mou_7)/2)

telecom_filtered['isd_ic_mou_diff'] = telecom_filtered.isd_ic_mou_8 - ((telecom_filtered.isd_ic_mou_6 + telecom_filtered.isd_ic_mou_7)/2)

telecom_filtered['spl_ic_mou_diff'] = telecom_filtered.spl_ic_mou_8 - ((telecom_filtered.spl_ic_mou_6 + telecom_filtered.spl_ic_mou_7)/2)

telecom_filtered['total_ic_mou_diff'] = telecom_filtered.total_ic_mou_8 - ((telecom_filtered.total_ic_mou_6 + telecom_filtered.total_ic_mou_7)/2)

telecom_filtered['total_rech_num_diff'] = telecom_filtered.total_rech_num_8 - ((telecom_filtered.total_rech_num_6 + telecom_filtered.total_rech_num_7)/2)

telecom_filtered['total_rech_amt_diff'] = telecom_filtered.total_rech_amt_8 - ((telecom_filtered.total_rech_amt_6 + telecom_filtered.total_rech_amt_7)/2)

telecom_filtered['max_rech_amt_diff'] = telecom_filtered.max_rech_amt_8 - ((telecom_filtered.max_rech_amt_6 + telecom_filtered.max_rech_amt_7)/2)

telecom_filtered['total_rech_data_diff'] = telecom_filtered.total_rech_data_8 - ((telecom_filtered.total_rech_data_6 + telecom_filtered.total_rech_data_7)/2)

telecom_filtered['max_rech_data_diff'] = telecom_filtered.max_rech_data_8 - ((telecom_filtered.max_rech_data_6 + telecom_filtered.max_rech_data_7)/2)

telecom_filtered['av_rech_amt_data_diff'] = telecom_filtered.av_rech_amt_data_8 - ((telecom_filtered.av_rech_amt_data_6 + telecom_filtered.av_rech_amt_data_7)/2)

telecom_filtered['vol_2g_mb_diff'] = telecom_filtered.vol_2g_mb_8 - ((telecom_filtered.vol_2g_mb_6 + telecom_filtered.vol_2g_mb_7)/2)

telecom_filtered['vol_3g_mb_diff'] = telecom_filtered.vol_3g_mb_8 - ((telecom_filtered.vol_3g_mb_6 + telecom_filtered.vol_3g_mb_7)/2)

## Data vis

### By looking at data , information is spread across 3 months - June,July and August. Hence creating new DF for months

In [25]:
def df_info(data) : 
    return pd.DataFrame({
                'Datatype' : data.dtypes.astype(str), 
                'Non_Null_Count': data.count(axis = 0).astype(int), 
                'Null_Count': data.isnull().sum().astype(int), 
                'Null_Percentage': round(data.isnull().sum()/len(data) * 100 , 2), 
                'Unique_Values_Count': data.nunique().astype(int) 
                 }).sort_values(by='Null_Percentage', ascending=False)

In [26]:
#June month
june_columns = []
for column in telecom.columns:
    x = re.search("6$", column)
    if x:
        june_columns.append(column)
# missing_values.loc[sixth_month_columns].sort_values(by='Null_Percentage', ascending=False)
june_info = df_info(telecom)
condition = june_info.index.isin(june_columns)
june_columns = june_info[condition]
june_columns

Datatype  Non_Null_Count  Null_Count  Null_Percentage  \
std_og_mou_6        float64           67231        2768             3.95   
loc_ic_t2m_mou_6    float64           67231        2768             3.95   
loc_ic_mou_6        float64           67231        2768             3.95   
isd_og_mou_6        float64           67231        2768             3.95   
std_ic_t2f_mou_6    float64           67231        2768             3.95   
og_others_6         float64           67231        2768             3.95   
std_ic_t2m_mou_6    float64           67231        2768             3.95   
std_ic_t2t_mou_6    float64           67231        2768             3.95   
spl_og_mou_6        float64           67231        2768             3.95   
loc_og_mou_6        float64           67231        2768             3.95   
std_og_t2f_mou_6    float64           67231        2768             3.95   
std_ic_mou_6        float64           67231        2768             3.95   
onnet_mou_6         float64           67231        2768             3.95   
loc_ic_t2f_mou_6    float64           67231        2768             3.95   
offnet_mou_6        float64           67231        2768             3.95   
ic_others_6         float64           67231        2768             3.95   
roam_ic_mou_6       float64           67231        2768             3.95   
roam_og_mou_6       float64           67231        2768             3.95   
loc_og_t2t_mou_6    float64           67231        2768             3.95   
isd_ic_mou_6        float64           67231        2768             3.95   
loc_og_t2m_mou_6    float64           67231        2768             3.95   
loc_og_t2f_mou_6    float64           67231        2768             3.95   
loc_og_t2c_mou_6    float64           67231        2768             3.95   
spl_ic_mou_6        float64           67231        2768             3.95   
std_og_t2t_mou_6    float64           67231        2768             3.95   
std_og_t2m_mou_6    float64           67231        2768             3.95   
loc_ic_t2t_mou_6    float64           67231        2768             3.95   
monthly_2g_6          int64           69999           0             0.00   
vol_3g_mb_6         float64           69999           0             0.00   
sachet_2g_6           int64           69999           0             0.00   
vol_2g_mb_6         float64           69999           0             0.00   
monthly_3g_6          int64           69999           0             0.00   
sachet_3g_6           int64           69999           0             0.00   
av_rech_amt_data_6  float64           69999           0             0.00   
arpu_6              float64           69999           0             0.00   
max_rech_data_6     float64           69999           0             0.00   
total_og_mou_6      float64           69999           0             0.00   
total_ic_mou_6      float64           69999           0             0.00   
total_rech_num_6      int64           69999           0             0.00   
total_rech_amt_6      int64           69999           0             0.00   
max_rech_amt_6        int64           69999           0             0.00   
last_day_rch_amt_6    int64           69999           0             0.00   
total_rech_data_6   float64           69999           0             0.00   

                    Unique_Values_Count  
std_og_mou_6                      21837  
loc_ic_t2m_mou_6                  19034  
loc_ic_mou_6                      24442  
isd_og_mou_6                       1103  
std_ic_t2f_mou_6                   2624  
og_others_6                         914  
std_ic_t2m_mou_6                   7931  
std_ic_t2t_mou_6                   5323  
spl_og_mou_6                       3503  
loc_og_mou_6                      22309  
std_og_t2f_mou_6                   2074  
std_ic_mou_6                       9876  
onnet_mou_6                       20057  
loc_ic_t2f_mou_6                   6200  
offnet_mou_6                      26254  
ic

In [27]:
#July month
july_columns = []
for column in telecom.columns:
    x = re.search("7$", column)
    if x:
        july_columns.append(column)
# missing_values.loc[sixth_month_columns].sort_values(by='Null_Percentage', ascending=False)
july_info = df_info(telecom)
condition = june_info.index.isin(july_columns)
july_columns = july_info[condition]
july_columns

Datatype  Non_Null_Count  Null_Count  Null_Percentage  \
loc_ic_mou_7        float64           67312        2687             3.84   
std_ic_t2t_mou_7    float64           67312        2687             3.84   
std_ic_t2m_mou_7    float64           67312        2687             3.84   
std_ic_t2f_mou_7    float64           67312        2687             3.84   
spl_ic_mou_7        float64           67312        2687             3.84   
isd_ic_mou_7        float64           67312        2687             3.84   
ic_others_7         float64           67312        2687             3.84   
std_ic_mou_7        float64           67312        2687             3.84   
spl_og_mou_7        float64           67312        2687             3.84   
loc_ic_t2f_mou_7    float64           67312        2687             3.84   
std_og_t2f_mou_7    float64           67312        2687             3.84   
loc_og_t2f_mou_7    float64           67312        2687             3.84   
loc_og_t2c_mou_7    float64           67312        2687             3.84   
roam_og_mou_7       float64           67312        2687             3.84   
roam_ic_mou_7       float64           67312        2687             3.84   
loc_ic_t2m_mou_7    float64           67312        2687             3.84   
loc_og_mou_7        float64           67312        2687             3.84   
std_og_t2t_mou_7    float64           67312        2687             3.84   
std_og_t2m_mou_7    float64           67312        2687             3.84   
std_og_mou_7        float64           67312        2687             3.84   
loc_og_t2t_mou_7    float64           67312        2687             3.84   
isd_og_mou_7        float64           67312        2687             3.84   
offnet_mou_7        float64           67312        2687             3.84   
og_others_7         float64           67312        2687             3.84   
onnet_mou_7         float64           67312        2687             3.84   
loc_ic_t2t_mou_7    float64           67312        2687             3.84   
loc_og_t2m_mou_7    float64           67312        2687             3.84   
av_rech_amt_data_7  float64           69999           0             0.00   
monthly_2g_7          int64           69999           0             0.00   
vol_3g_mb_7         float64           69999           0             0.00   
vol_2g_mb_7         float64           69999           0             0.00   
sachet_2g_7           int64           69999           0             0.00   
monthly_3g_7          int64           69999           0             0.00   
sachet_3g_7           int64           69999           0             0.00   
max_rech_data_7     float64           69999           0             0.00   
total_og_mou_7      float64           69999           0             0.00   
arpu_7              float64           69999           0             0.00   
total_ic_mou_7      float64           69999           0             0.00   
total_rech_num_7      int64           69999           0             0.00   
total_rech_amt_7      int64           69999           0             0.00   
max_rech_amt_7        int64           69999           0             0.00   
last_day_rch_amt_7    int64           69999           0             0.00   
total_rech_data_7   float64           69999           0             0.00   

                    Unique_Values_Count  
loc_ic_mou_7                      24258  
std_ic_t2t_mou_7                   5479  
std_ic_t2m_mou_7                   8033  
std_ic_t2f_mou_7                   2695  
spl_ic_mou_7                        103  
isd_ic_mou_7                       4640  
ic_others_7                        1658  
std_ic_mou_7                      10112  
spl_og_mou_7                       3827  
loc_ic_t2f_mou_7                   6287  
std_og_t2f_mou_7                   1991  
loc_og_t2f_mou_7                   3365  
loc_og_t2c_mou_7                   2103  
roam_og_mou_7                      5305  
roam_ic_mou_7                      4240  
lo

In [28]:
#August month
aug_columns = []
for column in telecom.columns:
    x = re.search("8$", column)
    if x:
        aug_columns.append(column)
# missing_values.loc[sixth_month_columns].sort_values(by='Null_Percentage', ascending=False)
aug_info = df_info(telecom)
condition = aug_info.index.isin(aug_columns)
aug_columns = july_info[condition]
aug_columns

Datatype  Non_Null_Count  Null_Count  Null_Percentage  \
std_og_t2m_mou_8    float64           66296        3703             5.29   
loc_og_mou_8        float64           66296        3703             5.29   
loc_ic_mou_8        float64           66296        3703             5.29   
loc_og_t2c_mou_8    float64           66296        3703             5.29   
std_og_mou_8        float64           66296        3703             5.29   
loc_og_t2f_mou_8    float64           66296        3703             5.29   
loc_ic_t2f_mou_8    float64           66296        3703             5.29   
isd_og_mou_8        float64           66296        3703             5.29   
loc_og_t2m_mou_8    float64           66296        3703             5.29   
spl_ic_mou_8        float64           66296        3703             5.29   
std_ic_t2f_mou_8    float64           66296        3703             5.29   
loc_og_t2t_mou_8    float64           66296        3703             5.29   
std_ic_t2m_mou_8    float64           66296        3703             5.29   
spl_og_mou_8        float64           66296        3703             5.29   
roam_og_mou_8       float64           66296        3703             5.29   
loc_ic_t2m_mou_8    float64           66296        3703             5.29   
std_og_t2t_mou_8    float64           66296        3703             5.29   
roam_ic_mou_8       float64           66296        3703             5.29   
isd_ic_mou_8        float64           66296        3703             5.29   
loc_ic_t2t_mou_8    float64           66296        3703             5.29   
offnet_mou_8        float64           66296        3703             5.29   
og_others_8         float64           66296        3703             5.29   
onnet_mou_8         float64           66296        3703             5.29   
std_ic_t2t_mou_8    float64           66296        3703             5.29   
std_og_t2f_mou_8    float64           66296        3703             5.29   
ic_others_8         float64           66296        3703             5.29   
std_ic_mou_8        float64           66296        3703             5.29   
vol_3g_mb_8         float64           69999           0             0.00   
av_rech_amt_data_8  float64           69999           0             0.00   
vol_2g_mb_8         float64           69999           0             0.00   
monthly_2g_8          int64           69999           0             0.00   
sachet_2g_8           int64           69999           0             0.00   
monthly_3g_8          int64           69999           0             0.00   
sachet_3g_8           int64           69999           0             0.00   
max_rech_data_8     float64           69999           0             0.00   
arpu_8              float64           69999           0             0.00   
total_og_mou_8      float64           69999           0             0.00   
total_ic_mou_8      float64           69999           0             0.00   
total_rech_num_8      int64           69999           0             0.00   
total_rech_amt_8      int64           69999           0             0.00   
max_rech_amt_8        int64           69999           0             0.00   
last_day_rch_amt_8    int64           69999           0             0.00   
total_rech_data_8   float64           69999           0             0.00   

                    Unique_Values_Count  
std_og_t2m_mou_8                  15968  
loc_og_mou_8                      21950  
loc_ic_mou_8                      24182  
loc_og_t2c_mou_8                   2204  
std_og_mou_8                      21745  
loc_og_t2f_mou_8                   3311  
loc_ic_t2f_mou_8                   6092  
isd_og_mou_8                       1008  
loc_og_t2m_mou_8                  17616  
spl_ic_mou_8                         95  
std_ic_t2f_mou_8                   2583  
loc_og_t2t_mou_8                  11296  
std_ic_t2m_mou_8                   7906  
spl_og_mou_8                       3848  
roam_og_mou_8                      5215  
lo